patroni/dcs/__init__.py

```diff
@@ -583,36 +583,63 @@ def use_slots(self) -> bool:
-        use_slots = self.use_slots
-        if role in ('master', 'primary', 'standby_leader'):
-            slot_members = [m.name for m in self.members if use_slots and m.name != my_name
-                            and (m.replicatefrom is None or m.replicatefrom == my_name
-                                 or not self.has_member(m.replicatefrom))]
-            permanent_slots = self.__permanent_slots if use_slots and \
-                role in ('master', 'primary') else self.__permanent_physical_slots
-        else:
-            # only manage slots for replicas that replicate from this one, except for the leader among them
-            slot_members = [m.name for m in self.members if use_slots
-                            and m.replicatefrom == my_name and m.name != self.leader_name]
-            permanent_slots = self.__permanent_logical_slots if use_slots and not nofailover else {}
+        """Lookup configured slot names in the DCS, report issues found and merge with permanent slots.
+
+        Will log an error if:
+
+            * Conflicting slot names between members are found
+            * Any logical slots are disabled, due to version compatibility, and *show_error* is ``True``.
+
+        :param my_name: name of this node.
+        :param role: role of this node.
+        :param nofailover: ``True`` if this node is tagged to not be a failover candidate.
+        :param major_version: postgresql major version.
+        :param show_error: if ``True`` report error if any disabled logical slots or conflicting slot names are found.
+
+        :returns: final dictionary of slot names, after merging with permanent slots and performing sanity checks.
+        """
+        slot_members: list[str] = self._get_slot_members(my_name, role)
 
-        slots = {slot_name_from_member_name(name): {'type': 'physical'} for name in slot_members}
+        slots: Dict[str, Dict[str, str]] = {slot_name_from_member_name(name): {'type': 'physical'}
+                                            for name in slot_members}
 
         if len(slots) < len(slot_members):
             # Find which names are conflicting for a nicer error message
             slot_conflicts: Dict[str, List[str]] = defaultdict(list)
             for name in slot_members:
                 slot_conflicts[slot_name_from_member_name(name)].append(name)
             logger.error("Following cluster members share a replication slot name: %s",
-                         "; ".join("{} map to {}".format(", ".join(v), k)
+                         "; ".join(f"{', '.join(v)} map to {k}"
                                    for k, v in slot_conflicts.items() if len(v) > 1))
 
-        # "merge" replication slots for members with permanent_replication_slots
+        disabled_permanent_logical_slots: list[str] = self._merge_permanent_slots(
+            slots, self._get_permanent_slots(role, nofailover), my_name, major_version)
```

Comment:
Similar to https://github.com/zalando/patroni/pull/2746/files#diff-247edd60473c0933e528a1b4621cc612da54b6dc964720719d7623135d312474R601:
```suggestion
        permanent_slots = self._get_permanent_slots(role, nofailover) if self.use_slots else {}
        disabled_permanent_logical_slots: list[str] = self._merge_permanent_slots(
            slots, permanent_slots, my_name, major_version)
```

and get rid of `use_slots` checks in the `_get_permanent_slots()`.

Reply:
Applied also in commit.

Reply:
You missed this part:
```python
        permanent_slots = self._get_permanent_slots(role, nofailover) if self.use_slots else {}
```

Reply:
Um, where? I don't see it?

Reply:
Currently it looks like:
```python
        disabled_permanent_logical_slots: list[str] = self._merge_permanent_slots(
            slots, self._get_permanent_slots(role, nofailover), my_name, major_version)
```

While it should be:
```python
        permanent_slots = self._get_permanent_slots(role, nofailover) if self.use_slots else {}
        disabled_permanent_logical_slots: list[str] = self._merge_permanent_slots(
            slots, permanent_slots, my_name, major_version)
```

Reply:
Oh duh! I was looking for something to remove when I should've been looking for something to add. :roll_eyes: 

Added that now. Sorry!